src/contexts/layout.tsx

```diff
@@ -1,94 +1,77 @@
+  fetchInitialConnections: () => void;
 }
 
-export const LayoutContext = createContext<ILayoutContext>(
-  {} as ILayoutContext,
-);
+export const useLayout = create<LayoutState>((set, get) => ({
+  keyspaces: {},
+  betterKeyspaces: {},
+  connections: [],
+  selectedConnection: undefined,
 
-export function LayoutProvider({ children }: { children: React.ReactNode }) {
-  const [keyspaces, setKeyspaces] = useState({});
-  const [betterKeyspaces, setBetterKeyspaces] = useState({});
-  const [connections, setConnections] = useState<Connection[]>([]);
-  const [selectedConnection, setSelectedConnection] = useState<Connection>();
+  setKeyspaces: (keyspaces) => set({ keyspaces }),
 
-  useEffect(() => {
-    const fetchInitialConnections = async () => {
-      const initialConnections = await fetchConnections();
-      startTransition(() => {
-        setConnections(initialConnections);
-      });
-    };
+  setBetterKeyspaces: (betterKeyspaces) => set({ betterKeyspaces }),
 
-    fetchInitialConnections();
-  }, []);
+  setConnections: (connections) => {
+    if (JSON.stringify(get().connections) !== JSON.stringify(connections)) {
+      set({ connections });
+    }
+  },
 
-  const queryKeyspace = useAction(queryKeyspaceAction, {
-    onExecute: () => {},
-    onSuccess: ({ data }) => {
-      console.log(data);
-      if (data?.keyspaces) setKeyspaces(data.keyspaces);
-      if (data?.betterKeyspaces) setBetterKeyspaces(data.betterKeyspaces);
-    },
-    onError: (error) => {
-      console.log(error);
-      toast.error("Failed to query keyspaces.");
-      setSelectedConnection(undefined);
-      setKeyspaces({});
-    },
-  });
+  setSelectedConnection: (connection) => {
+    set({ selectedConnection: connection });
+  },
 
-  useEffect(() => {
-    if (selectedConnection) {
-      queryKeyspace.execute({
+  fetchInitialConnections: async () => {
+    try {
+      const initialConnections = await fetchConnections();
+      if (
+        JSON.stringify(get().connections) !== JSON.stringify(initialConnections)
+      ) {
+        startTransition(() => {
+          set({ connections: initialConnections });
+        });
+      }
+    } catch (error) {
+      console.error("Error fetching connections:", error);
+      toast.error("Failed to fetch connections.");
+    }
+  },
+
+  queryKeyspace: async (selectedConnection) => {
+    try {
+      const queryKeyspace = await queryKeyspaceAction({
         host: selectedConnection.host,
         port: selectedConnection.port,
         password: selectedConnection.password,
         username: selectedConnection.username,
       });
+      return queryKeyspace;
+    } catch (error) {
+      console.error(error);
+      toast.error("Failed to query keyspaces.");
+      get().setSelectedConnection(undefined);
+      throw error;
     }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [queryKeyspace.execute, selectedConnection]);
```

Comment:
The problem of removing the `useEffect` is that the keyspaces are not fetched anymore. Neither the initial connections (the `useEffect` above)

Reply:
@Matozinho the problem I was facing with hooks was that since it is not a component and is zustand store we can't use hook.

Reply:
Sure will let you know!